File: examples/plugins/js/mangzio/enrich.js
// enrich.js — generic alt-title/alt-summary enrichment providers (JS).
// Copy this file into any JS plugin folder unchanged and require() it from
// main.js BEFORE the PLUGIN declaration, e.g.:
//   require("./enrich.js");
// Then declare the servers you want in PLUGIN.alt_title_servers:
//   alt_title_servers: [
//       { id: "mangadex", name: "MangaDex", kind: "titles" },
//       { id: "mangaupdates", name: "MangaUpdates", kind: "both" },
//   ],
// kind: "titles" (default) = appears in the alt-title picker only;
//       "summaries"         = alt-summary picker only;
//       "both"              = shows up in both pickers.
//
// Self-contained: only the sandbox global http_request is used (no httpGet,
// no per-site globals like API_URL — enrichment hits external APIs directly).
// require() runs this file into the plugin's own goja VM, so the functions
// below become globals; strip any inline getAltTitles/getAltSummary from
// your main.js before requiring this file (a later definition would win).
//
// ABI exports provided:
//   getAltTitles(arg)   arg {"title","server"} -> {source, titles[]}
//   getAltSummary(arg)  arg {"title","server"} -> {source, summaries[]}
//     server values handled: "mangadex", "mangaupdates"
//     unknown server -> empty result for the requested kind

// Local GET/POST over the sandbox http_request global.
function enrichHttp(url, opts) {
    var method = (opts && opts.method) || "GET";
    var headers = (opts && opts.headers) || {};
    var body = (opts && opts.body) || "";
    var resp = http_request(JSON.stringify({
        method: method,
        url: url,
        headers: headers,
        body: body
    }));
    if (typeof resp === "string") {
        try { resp = JSON.parse(resp); } catch (e) { return null; }
    }
    return resp;
}

// Search the MangaUpdates v1 API by title; returns best-match record or null.
function mangaupdatesSearch(title) {
    var resp = enrichHttp("https://api.mangaupdates.com/v1/series/search", {
        method: "POST",
        headers: { "Content-Type": "application/json", "Accept": "application/json" },
        body: JSON.stringify({ search: title, stype: "title", perpage: 5 })
    });
    if (!resp || resp.status !== 200) return null;
    try {
        var body = JSON.parse(resp.body);
        if (!body || !body.results || body.results.length === 0) return null;
        return body.results[0].record; // API sorts by relevance
    } catch (e) { return null; }
}

// Fetch the full MangaUpdates series record (has associated[] alt titles).
function mangaupdatesDetail(seriesId) {
    var resp = enrichHttp("https://api.mangaupdates.com/v1/series/" + seriesId);
    if (!resp || resp.status !== 200) return null;
    try { return JSON.parse(resp.body); } catch (e) { return null; }
}

function mangadexAltTitles(title) {
    var qs = "title=" + encodeURIComponent(title) + "&limit=5&includes[]=manga";
    var resp = enrichHttp("https://api.mangadex.org/manga?" + qs);
    if (!resp || resp.status < 200 || resp.status >= 300) {
        return JSON.stringify({ source: "MangaDex", titles: [] });
    }
    var body;
    try { body = JSON.parse(resp.body); } catch (e) {
        return JSON.stringify({ source: "MangaDex", titles: [] });
    }
    var data = body && body.data;
    if (!data || data.length === 0) {
        return JSON.stringify({ source: "MangaDex", titles: [] });
    }
    var attrs = data[0].attributes;
    var out = [], seen = {};
    var keepLangs = { "en": true, "ja": true, "ja-ro": true, "ko": true, "ko-ro": true };
    if (attrs && attrs.altTitles) {
        for (var i = 0; i < attrs.altTitles.length; i++) {
            var alt = attrs.altTitles[i];
            var keys = Object.keys(alt);
            for (var j = 0; j < keys.length; j++) {
                var lang = keys[j];
                if (keepLangs[lang] && alt[lang] && !seen[alt[lang]]) {
                    seen[alt[lang]] = true;
                    out.push(alt[lang]);
                }
            }
        }
    }
    return JSON.stringify({ source: "MangaDex", titles: out });
}

function mangaupdatesAltTitles(title) {
    var record = mangaupdatesSearch(title);
    if (!record) return JSON.stringify({ source: "MangaUpdates", titles: [] });
    var detail = mangaupdatesDetail(record.series_id);
    if (!detail || !detail.associated) {
        return JSON.stringify({ source: "MangaUpdates", titles: [] });
    }
    var out = [], seen = {};
    for (var i = 0; i < detail.associated.length; i++) {
        var t = detail.associated[i].title;
        if (t && !seen[t]) { seen[t] = true; out.push(t); }
    }
    return JSON.stringify({ source: "MangaUpdates", titles: out });
}

// getAltTitles(arg {"title","server"}) -> {"source", "titles":[]}
function getAltTitles(arg) {
    var input = JSON.parse(arg);
    var title = input.title || "";
    var server = input.server || "mangadex";

    if (server === "mangaupdates") {
        return mangaupdatesAltTitles(title);
    }
    // default: MangaDex
    return mangadexAltTitles(title);
}

// getAltSummary(arg {"title","server"}) -> {"source", "summaries":[]}
// Only MangaUpdates is a summary provider today.
function getAltSummary(arg) {
    var input = JSON.parse(arg);
    var title = input.title || "";
    var server = input.server || "mangaupdates";

    if (server !== "mangaupdates") {
        return JSON.stringify({ source: server, summaries: [] });
    }
    var record = mangaupdatesSearch(title);
    if (!record) return JSON.stringify({ source: "MangaUpdates", summaries: [] });
    var desc = record.description || "";
    if (!desc) {
        var detail = mangaupdatesDetail(record.series_id);
        if (detail) desc = detail.description || "";
    }
    if (!desc) return JSON.stringify({ source: "MangaUpdates", summaries: [] });
    return JSON.stringify({ source: "MangaUpdates", summaries: [desc] });
}

File: examples/plugins/lua/mangabuddy/enrich.lua
-- enrich.lua — generic alt-title/alt-summary enrichment providers.
-- Copy this file into any plugin folder unchanged and declare the servers you
-- want in your PLUGIN table, e.g.:
--   alt_title_servers = {
--       {id = "mangadex",     name = "MangaDex",     kind = "titles"},
--       {id = "mangaupdates", name = "MangaUpdates", kind = "both"}
--   },
-- kind: "titles" (default) = appears in the alt-title picker only;
--       "summaries"         = alt-summary picker only;
--       "both"              = shows up in both pickers.
-- Requires the sandbox globals http_request / json / log and url_encode from
-- helpers.lua (already present in this folder's split). Sibling modules
-- pre-execute before main.lua, so the ABI exports below are visible at call
-- time regardless of file order.
--
-- ABI exports provided:
--   getAltTitles(arg)   arg {"title","server"} -> {source, titles[]}
--   getAltSummary(arg)  arg {"title","server"} -> {source, summaries[]}
--     server values handled: "mangadex", "mangaupdates"
--     unknown server -> empty result for the requested kind

-- ─── MangaUpdates helpers (local to this module) ───────────────────────────

-- search the MangaUpdates v1 API by title; returns best-match record or nil.
local function mangaupdates_search(title)
    local body = json.encode({search = title, stype = "title", perpage = 5})
    local resp = http_request({
        url = "https://api.mangaupdates.com/v1/series/search",
        method = "POST",
        headers = {
            ["Content-Type"] = "application/json",
            ["Accept"] = "application/json"
        },
        body = body
    })
    if not resp or resp.status ~= 200 then
        log.error("mangaupdates search http " .. tostring(resp and resp.status or "nil"))
        return nil
    end
    local ok, data = pcall(json.decode, resp.body)
    if not ok or not data or not data.results or #data.results == 0 then
        return nil
    end
    -- best match = first result (API sorts by relevance)
    return data.results[1].record
end

local function mangaupdates_detail(series_id)
    local resp = http_request({
        url = "https://api.mangaupdates.com/v1/series/" .. tostring(series_id),
        method = "GET",
        headers = {["Accept"] = "application/json"}
    })
    if not resp or resp.status ~= 200 then
        log.error("mangaupdates detail http " .. tostring(resp and resp.status or "nil"))
        return nil
    end
    local ok, data = pcall(json.decode, resp.body)
    if not ok or not data then return nil end
    return data
end

-- ─── MangaDex alt titles ───────────────────────────────────────────────────

local function altTitles_mangadex(title)
    local url = "https://api.mangadex.org/manga?title=" .. url_encode(title) ..
        "&limit=5&includes[]=manga"
    local resp = http_request({url = url, method = "GET", headers = {}})
    if not resp or resp.status ~= 200 then
        return json.encode({source = "MangaDex", titles = {}})
    end
    local ok, body = pcall(json.decode, resp.body)
    if not ok or not body or not body.data or #body.data == 0 then
        return json.encode({source = "MangaDex", titles = {}})
    end
    local attrs = body.data[1].attributes
    local out = {}
    local seen = {}
    local keep_langs = {en = true, ja = true, ["ja-ro"] = true, ko = true, ["ko-ro"] = true}
    if attrs.altTitles then
        for _, alt in ipairs(attrs.altTitles) do
            for lang, val in pairs(alt) do
                if keep_langs[lang] and val and not seen[val] then
                    seen[val] = true
                    out[#out + 1] = val
                end
            end
        end
    end
    return json.encode({source = "MangaDex", titles = out})
end

-- ─── MangaUpdates alt titles ───────────────────────────────────────────────

local function altTitles_mangaupdates(title)
    local record = mangaupdates_search(title)
    if not record then
        return json.encode({source = "MangaUpdates", titles = {}})
    end
    local detail = mangaupdates_detail(record.series_id)
    if not detail or not detail.associated then
        return json.encode({source = "MangaUpdates", titles = {}})
    end
    local out = {}
    local seen = {}
    for _, item in ipairs(detail.associated) do
        if item.title and item.title ~= "" and not seen[item.title] then
            seen[item.title] = true
            out[#out + 1] = item.title
        end
    end
    return json.encode({source = "MangaUpdates", titles = out})
end

-- ─── MangaUpdates alt summary ──────────────────────────────────────────────

local function altSummary_mangaupdates(title)
    local record = mangaupdates_search(title)
    if not record then
        return json.encode({source = "MangaUpdates", summaries = {}})
    end
    -- The search response already includes the description in the record.
    local desc = record.description or ""
    if desc == "" then
        -- Fallback: fetch full detail.
        local detail = mangaupdates_detail(record.series_id)
        if detail then desc = detail.description or "" end
    end
    if desc == "" then
        return json.encode({source = "MangaUpdates", summaries = {}})
    end
    return json.encode({source = "MangaUpdates", summaries = {desc}})
end

-- ─── ABI dispatch ──────────────────────────────────────────────────────────

function getAltTitles(arg)
    local input = json.decode(arg)
    local title = input.title or ""
    local server = input.server or "mangadex"

    if server == "mangaupdates" then
        return altTitles_mangaupdates(title)
    end
    -- default: MangaDex
    return altTitles_mangadex(title)
end

function getAltSummary(arg)
    local input = json.decode(arg)
    local title = input.title or ""
    local server = input.server or "mangaupdates"

    if server == "mangaupdates" then
        return altSummary_mangaupdates(title)
    end
    -- No other summary provider known; return empty.
    return json.encode({source = server, summaries = {}})
end
